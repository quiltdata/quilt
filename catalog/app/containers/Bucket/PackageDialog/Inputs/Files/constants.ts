import cfg from 'constants/config'

export const MAX_UPLOAD_SIZE = 20 * 1000 * 1000 * 1000 // 20GB
// XXX: keep in sync w/ the backend
// NOTE: these limits are lower than the actual "hard" limits on the backend
export const MAX_S3_SIZE = cfg.chunkedChecksums
  ? 5 * 10 ** 12 // 5 TB
  : 50 * 10 ** 9 // 50 GB
export const MAX_FILE_COUNT = 1000
