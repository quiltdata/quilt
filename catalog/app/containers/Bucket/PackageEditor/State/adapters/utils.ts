import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
