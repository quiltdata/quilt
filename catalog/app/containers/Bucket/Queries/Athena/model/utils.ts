export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
