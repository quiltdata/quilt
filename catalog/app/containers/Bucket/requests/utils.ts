import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
