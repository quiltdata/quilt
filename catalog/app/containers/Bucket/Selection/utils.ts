import { join } from 'path'

import * as R from 'ramda'

import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export interface SelectionItem {
  logicalKey: string
}

export interface ListingSelection {
  [prefixUrl: string]: SelectionItem[]
}

export const EMPTY_MAP: ListingSelection = {}

interface SelectionHandles {
  [prefixUrl: string]: Model.S3.S3ObjectLocation[]
}

const convertIdToHandle = (
  id: string | number,
  parentHandle: Model.S3.S3ObjectLocation,
): Model.S3.S3ObjectLocation => ({
  bucket: parentHandle.bucket,
  key: join(parentHandle.key, id.toString()),
})

export const toHandlesMap = (selection: ListingSelection): SelectionHandles =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => ({
      ...memo,
      [prefixUrl]: items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    }),
    {} as SelectionHandles,
  )

export const toHandlesList = (selection: ListingSelection): Model.S3.S3ObjectLocation[] =>
  Object.entries(selection).reduce(
    (memo, [prefixUrl, items]) => [
      ...memo,
      ...items.map((item) =>
        convertIdToHandle(item.logicalKey, s3paths.parseS3Url(prefixUrl)),
      ),
    ],
    [] as Model.S3.S3ObjectLocation[],
  )

const mergeWithFiltered =
  (prefix: string, filteredItems: SelectionItem[]) => (allItems: SelectionItem[]) => {
    if (!allItems || !allItems.length) return filteredItems
    const selectionOutsideFilter = allItems.filter(
      (item) => !item.logicalKey.startsWith(prefix),
    )
    const newIds = [...selectionOutsideFilter, ...filteredItems]
    return R.equals(newIds, allItems) ? allItems : newIds // avoids cyclic update
  }

export function merge(
  items: SelectionItem[],
  bucket: string,
  path: string,
  filter?: string,
): (state: ListingSelection) => ListingSelection {
  const prefixUrl = `s3://${bucket}/${path}`
  const lens = R.lensProp<Record<string, SelectionItem[]>>(prefixUrl)
  return filter ? R.over(lens, mergeWithFiltered(filter, items)) : R.set(lens, items)
}

const EmptyKeys: string[] = []

export const getDirectorySelection = (
  selection: ListingSelection,
  bucket: string,
  path: string,
) => selection[`s3://${bucket}/${path}`] || EmptyKeys
