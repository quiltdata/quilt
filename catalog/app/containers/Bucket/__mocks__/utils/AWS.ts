import { handleToS3Url } from 'utils/s3paths'

export const Signer = {
  useDownloadUrl: handleToS3Url,
}
