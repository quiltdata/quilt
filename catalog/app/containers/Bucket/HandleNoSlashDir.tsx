import * as React from 'react'
import * as RRDom from 'react-router-dom'

import Placeholder from 'components/Placeholder'
import * as Model from 'model'
import * as AWS from 'utils/AWS'
import * as NamedRoutes from 'utils/NamedRoutes'
import * as s3paths from 'utils/s3paths'

import * as requests from './requests'
import { displayError } from './errors'

// If object exists, then this is 100% an object page
function useIsObject(handle: Model.S3.S3ObjectLocation) {
  const [exists, setExists] = React.useState(null)

  const s3 = AWS.S3.use()
  React.useEffect(() => {
    const { bucket, key, version } = handle
    function resolveObjectExistence() {
      requests
        .getObjectExistence({
          s3,
          bucket,
          key,
          version,
        })
        .then(
          requests.ObjectExistence.case({
            Exists: () => true,
            _: () => false,
          }),
        )
        .then(setExists)
    }
    resolveObjectExistence()
  }, [s3, handle])

  return exists
}

// If prefix contains at least something, then it is a directory.
// S3 can not have empty directories, because directories are virtual,
//    they are based on paths for existing keys (file)
function useIsDirectory(handle: Model.S3.S3ObjectLocation) {
  const bucketListing = requests.useBucketListing()
  return React.useCallback(async () => {
    const { bucket, key } = handle
    const path = s3paths.ensureSlash(key)
    const { dirs, files } = await bucketListing({ bucket, path, maxKeys: 1 })
    return !!dirs.length || !!files.length
  }, [bucketListing, handle])
}

function useHandleNoSlashDir(
  handle: Model.S3.S3ObjectLocation,
): 'loading' | 'dir' | 'file' | Error {
  const isObject = useIsObject(handle)
  const requestIsDirectory = useIsDirectory(handle)

  const [pageType, setPageType] = React.useState<'loading' | 'dir' | 'file' | Error>(
    'loading',
  )

  React.useEffect(() => {
    function fetchData() {
      if (isObject === null) return

      if (isObject) {
        setPageType('file')
        return
      }

      requestIsDirectory()
        .then((isDir) => setPageType(isDir ? 'dir' : 'file'))
        .catch((e) => setPageType(e instanceof Error ? e : new Error(`${e}`)))
    }
    fetchData()
  }, [handle, isObject, requestIsDirectory])

  return pageType
}

interface HandleNoSlashDirProps {
  children: React.ReactNode
  handle: Model.S3.S3ObjectLocation
}

export default function HandleNoSlashDir({ children, handle }: HandleNoSlashDirProps) {
  const { urls } = NamedRoutes.use()

  const pageType = useHandleNoSlashDir(handle)

  switch (pageType) {
    case 'loading':
      return <Placeholder color="text.secondary" />
    case 'dir':
      const dirPage = urls.bucketDir(handle.bucket, s3paths.ensureSlash(handle.key))
      return <RRDom.Redirect to={dirPage} />
    case 'file':
      return children
    default:
      displayError()(pageType)
  }
}
