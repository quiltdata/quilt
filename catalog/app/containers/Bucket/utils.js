import * as React from 'react'

import * as AWS from 'utils/AWS'

export const withDownloadUrl = (handle, callback) => (
  <AWS.Signer.Inject>
    {(signer) =>
      callback(
        signer.getSignedS3URL(handle, { ResponseContentDisposition: 'attachment' }),
      )
    }
  </AWS.Signer.Inject>
)
