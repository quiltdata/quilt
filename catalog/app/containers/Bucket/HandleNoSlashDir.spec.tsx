import * as React from 'react'
import { act, create } from 'react-test-renderer'

import { bucketDir, bucketPackageTree } from 'constants/routes'
import * as NamedRoutes from 'utils/NamedRoutes'

import HandleNoSlashDir from './HandleNoSlashDir'

jest.mock(
  'constants/config',
  jest.fn(() => ({})),
)

jest.mock('components/Placeholder', () => () => <h1>Loadingâ€¦</h1>)

const headObjectPromise = jest.fn(() => ({}))

const listObjectsV2Promise = jest.fn(() => ({
  CommonPrefixes: [{ Prefix: 'foo' }],
}))

jest.mock('utils/AWS', () => ({
  S3: {
    use: () => ({
      headObject: jest.fn(() => ({
        response: {
          httpResponse: {
            headers: {},
          },
        },
        promise: headObjectPromise,
      })),
      listObjectsV2: jest.fn(() => ({
        promise: listObjectsV2Promise,
      })),
    }),
  },
}))

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  Redirect: () => 'redirect',
}))

function wait(timeout: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve()
    }, timeout)
  })
}

describe('containers/Bucket/HandleNoSlashDir', () => {
  const handle = { bucket: 'b', key: 'k', version: 'some' }

  function TestWrapper() {
    return (
      <NamedRoutes.Provider routes={{ bucketDir, bucketPackageTree }}>
        <HandleNoSlashDir handle={handle}>
          <h1>It works</h1>
        </HandleNoSlashDir>
      </NamedRoutes.Provider>
    )
  }

  it('renders placeholder', async () => {
    const tree = create(<TestWrapper />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders file content', async () => {
    const tree = create(<TestWrapper />)
    await act(() => wait(100))
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('redirects', async () => {
    headObjectPromise.mockImplementation(() => {
      throw {
        code: 'NotFound',
      }
    })
    const tree = create(<TestWrapper />)
    await act(() => wait(100)) // object request and state change
    await act(() => wait(100)) // listing request and state change
    expect(tree.toJSON()).toMatchSnapshot()
  })
})
