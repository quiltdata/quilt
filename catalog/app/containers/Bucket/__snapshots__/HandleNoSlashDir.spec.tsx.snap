// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/HandleNoSlashDir redirects 1`] = `"redirect"`;

exports[`containers/Bucket/HandleNoSlashDir renders file content 1`] = `
<h1>
  It works
</h1>
`;

exports[`containers/Bucket/HandleNoSlashDir renders placeholder 1`] = `
<h1>
  Loadingâ€¦
</h1>
`;
